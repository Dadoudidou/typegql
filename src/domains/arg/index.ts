import { argRegistry } from './registry';
export { compileFieldArgs } from './compiler';
import { ArgOptions, defaultArgOptions } from './options';

export function Arg(options: ArgOptions = {}): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    const compiledOptions = {
      ...defaultArgOptions,
      ...options,
    }
    argRegistry.set(target.constructor, [fieldName, argIndex], compiledOptions);
  };
}
