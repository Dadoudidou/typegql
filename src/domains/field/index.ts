import { fieldsRegistry, FieldInnerConfig } from './registry';

export { FieldInnerConfig, fieldsRegistry, queryFieldsRegistry } from './registry';
export { compileAllFields, compileFieldConfig } from './compiler';
export { isQueryField } from './services';
export { FieldError } from './error';

export interface FieldOptions {
  description?: string;
  type?: any;
  name?: string;
}

export function Field(options?: FieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      ...options,
    };

    fieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig);
  };
}
