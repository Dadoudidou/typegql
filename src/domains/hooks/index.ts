import {
  registerFieldAfterHook,
  registerFieldBeforeHook,
  HookExecutor,
} from './registry';

export {
  fieldAfterHooksRegistry,
  fieldBeforeHooksRegistry,
  HookExecutor,
} from './registry';
export { HookError } from './error';

export function Before(hook: HookExecutor): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    registerFieldBeforeHook(targetInstance.constructor, fieldName, hook);
  };
}

export function After(hook: HookExecutor): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    registerFieldAfterHook(targetInstance.constructor, fieldName, hook);
  };
}

export interface GuardOptions {
  msg: string;
}

export function Guard(
  guardFunction: HookExecutor<boolean>,
  options: GuardOptions,
): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const hook: HookExecutor = async data => {
      const isAllowed = await guardFunction(data);
      if (isAllowed !== true) {
        throw new Error(options.msg);
      }
    };
    registerFieldBeforeHook(targetInstance.constructor, fieldName, hook);
  };
}

export function createGuard(guardFunction: HookExecutor<boolean>, options: GuardOptions) {
  return (guardOptions: GuardOptions = options): PropertyDecorator => {
    return Guard(guardFunction, guardOptions);
  };
}
