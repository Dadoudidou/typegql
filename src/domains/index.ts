export { Type, compileType } from './type';
export { Field } from './field';
export { Arg } from './arg';
export { Inject, Context, Source, Info } from './inject';
export { registerEnum } from './enum';
