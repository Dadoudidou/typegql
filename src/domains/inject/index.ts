import { InjectorResolver, injectorRegistry } from './registry';
export { injectorRegistry, InjectorsIndex, InjectorResolver } from './registry';

export function Inject(resolver: InjectorResolver): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    injectorRegistry.set(target.constructor, [fieldName, argIndex], resolver);
  };
}

export const Context = Inject((source, args, context, info) => {
  return context;
});

export const Info = Inject((source, args, context, info) => {
  return info;
});

export const Source = Inject((source, args, context, info) => {
  return source;
});
