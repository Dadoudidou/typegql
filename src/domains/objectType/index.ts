import { compileObjectTypeWithConfig, compileObjectTypeAsInputWithConfig } from './compiler';
import { objectTypeRegistry, inputTypeRegistry } from './registry';

export { compileObjectType } from './compiler';
export { ObjectTypeError } from './error';
export { objectTypeRegistry, inputTypeRegistry } from './registry';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, config);
    objectTypeRegistry.set(target, outputTypeCompiler);
    
    const inputTypeCompiler = () =>
    compileObjectTypeAsInputWithConfig(target, config);
    inputTypeRegistry.set(target, inputTypeCompiler);
  };
}
