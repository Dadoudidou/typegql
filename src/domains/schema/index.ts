export { schemaRegistry, mutationFieldsRegistry, queryFieldsRegistry } from './registry';
import { schemaRegistry } from './registry';
import { compileSchema } from './compiler';
export { compileSchema } from './compiler';
export { Query, Mutation } from './rootFields';

export function Schema(): ClassDecorator {
  return target => {
    const compiler = () => compileSchema(target);
    schemaRegistry.set(target, compiler);
  };
}
