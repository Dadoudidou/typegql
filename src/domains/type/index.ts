import { registerType } from './registry';

export { objectTypeRegistry, inputTypeRegistry, registerType } from './registry';
export { ObjectTypeError } from './error'

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    registerType(target, config);
  };
}
