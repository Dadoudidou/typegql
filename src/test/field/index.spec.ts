import { GraphQLString, GraphQLFloat, GraphQLBoolean } from 'graphql';
import { Type, Field, compileType } from 'domains';

import 'reflect-metadata';

describe('Field', () => {
  it('Resolves fields with default value', async () => {
    @Type()
    class Foo {
      @Field() bar: string = 'baz';
    }
    const compiled = compileType(Foo);
    const barField = compiled.getFields().bar;
    expect(await barField.resolve(new Foo(), {}, null, null)).toEqual('baz');
  });

  it('Resolves fields with function resolver', async () => {
    @Type()
    class Foo {
      @Field()
      bar(): string {
        return 'baz';
      }
    }

    const compiled = compileType(Foo);
    const barField = compiled.getFields().bar;

    expect(await barField.resolve(new Foo(), {}, null, null as any)).toEqual('baz');
  });

  it('Handles description', () => {
    @Type()
    class Foo {
      @Field({ description: 'test' })
      bar: string = 'baz';
    }
    expect(compileType(Foo).getFields().bar.description).toEqual('test');
  });

  it('Handles custom name', async () => {
    @Type()
    class Foo {
      @Field({ name: 'baz', description: 'test' })
      bar: string = 'test';
    }
    const compiled = compileType(Foo);
    const bazField = compiled.getFields().baz;
    expect(compiled.getFields().bar).toBeFalsy();
    expect(bazField).toBeTruthy();
    expect(bazField.description).toEqual('test');
    expect(await bazField.resolve(new Foo(), {}, null, null as any)).toBe('test');
  });

  it('Properly infers basic scalar types', () => {
    @Type()
    class Foo {
      @Field() bar: string;
      @Field() baz: number;
      @Field() foo: boolean;
      @Field() coo: boolean = false;
      @Field()
      boo(): boolean {
        return true;
      }
    }

    const { bar, baz, foo, boo, coo } = compileType(Foo).getFields();

    expect(bar.type).toEqual(GraphQLString);
    expect(baz.type).toEqual(GraphQLFloat);
    expect(foo.type).toEqual(GraphQLBoolean);
    expect(boo.type).toEqual(GraphQLBoolean);
    expect(coo.type).toEqual(GraphQLBoolean);
  });

  it('Properly sets forced field type', () => {
    @Type()
    class Foo {
      @Field({ type: () => GraphQLFloat })
      bar: string;
    }

    const { bar } = compileType(Foo).getFields();
    expect(bar.type).toEqual(GraphQLFloat);
  });

  it('Supports references to other types', () => {
    @Type()
    class Foo {
      @Field() foo: string;
    }

    @Type()
    class Bar {
      @Field() foo: Foo;
    }

    const { foo } = compileType(Bar).getFields();
    const compiledFoo = compileType(Foo);
    expect(foo.type).toBe(compiledFoo);
  });

  it('Supports references to itself', () => {
    @Type()
    class Foo {
      @Field() fooNested: Foo;
    }

    const { fooNested } = compileType(Foo).getFields();
    expect(fooNested.type).toBe(compileType(Foo));
  });

  it('Supports circular references', () => {
    @Type()
    class Car {
      @Field({ type: () => Owner })
      owner: Owner;
    }

    @Type()
    class Owner {
      @Field({ type: () => Car })
      car: Car;
    }

    // console.log({ Car, Owner });

    const { owner } = compileType(Car).getFields();
    const { car } = compileType(Owner).getFields();

    // console.log({ owner, car });
    expect(owner.type).toBe(compileType(Owner));
    expect(car.type).toBe(compileType(Car));
  });

  it('Throws if pointing to unregistered type', () => {
    class Foo {}

    @Type()
    class Bar {
      @Field({ type: () => Foo })
      foo: Foo;
    }

    expect(() => compileType(Bar).getFields()).toThrowErrorMatchingSnapshot();
  });
});
