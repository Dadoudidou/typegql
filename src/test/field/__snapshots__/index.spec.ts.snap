// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Throws if pointing to unregistered type 1`] = `"@Type Bar.foo: Forced type is incorrect. Make sure to use either native graphql type or class that is registered with @Type decorator"`;
